src/AvifLoad.cs

```diff
@@ -141,5 +162,148 @@ private static void AddAvifMetadataToDocument(Document doc, AvifReader reader)
+                {
+                    converted.CopyPixels(output.AsRegionPtr().Cast<ColorBgra32>());
+                }
+            }
+            else
+            {
+                using (IBitmapSource<ColorBgra32> converted = image.Image.CreateFormatConverter<ColorBgra32>())
+                {
+                    converted.CopyPixels(output.AsRegionPtr().Cast<ColorBgra32>());
+                }
+            }
+        }
+
+        private static void SetOutputLayerDataRgba128PQ(
+            Document document,
+            AvifReaderImage image,
+            Surface output,
+            IImagingFactory imagingFactory)
+        {
+            using (IColorContext dp3ColorContext = imagingFactory.CreateColorContext(KnownColorSpace.DisplayP3))
+            using (IDirect2DFactory d2dFactory = Direct2DFactory.Create())
+            {
+                if (image.IsPremultipliedAlpha)
+                {
+                    using (IBitmapSource<ColorPbgra32> dp3Image = PQToDisplayP3(image.Image,
+                                                                                ColorManagementAlphaMode.Premultiplied,
+                                                                                imagingFactory,
+                                                                                d2dFactory,
+                                                                                dp3ColorContext))
+                    {
+                        dp3Image.CopyPixels(output.AsRegionPtr().Cast<ColorPbgra32>());
+                    }
+                }
+                else
+                {
+                    // Direct2D requires everything to be "premultiplied"
+                    using (IBitmap asPrgba128Float = image.Image.Cast<ColorPrgba128Float>())
+                    using (IBitmapSource<ColorPbgra32> dp3Image = PQToDisplayP3(asPrgba128Float,
+                                                                                ColorManagementAlphaMode.Straight,
+                                                                                imagingFactory,
+                                                                                d2dFactory,
+                                                                                dp3ColorContext))
+                    {
+                        dp3Image.CopyPixels(output.AsRegionPtr().Cast<ColorPbgra32>());
+                    }
+                }
+
+                document.SetColorContext(dp3ColorContext);
+            }
+
+            static IBitmapSource<ColorPbgra32> PQToDisplayP3(
+                IBitmap bitmap,
+                ColorManagementAlphaMode alphaMode,
+                IImagingFactory imagingFactory,
+                IDirect2DFactory d2dFactory,
+                IColorContext dp3ColorContext)
+            {
+                return d2dFactory.CreateBitmapSourceFromImage<ColorPbgra32>(
```

Comment:
Your code here is fine, but I did want to point out that it's important to be careful with the `IBitmapSource` you get from `CreateBitmapSourceFromImage`. Each call into its `CopyPixels` will result in a Direct2D `BeginDraw`+`EndDraw` sequence, and there's a good amount of minimum overhead in each `Begin+EndDraw` pass. So you should be careful to do bulk processing, and _not_ row-at-a-time processing. Many WIC transforms will do row-at-a-time processing, e.g. format converters and the like, because they copy into a temporary buffer. So if you find you're getting suuuuuuuuper slow performance, use `.ToBitmap()` and then do processing on that.